### DISCUSSION TITLE
Allow Dynamic Import in `commonjs2`

### DISCUSSION THREAD

**Discussion by neurolag** [webpack/webpack#16272] [Q&A] [answered]
## Feature request

**What is the expected behavior?**
webpack should allow `module` dependencies to be declared using the `externals` setting.

Furthermore, dynamic `import()` calls should be preserved.

**What is motivation or use case for adding/changing the behavior?**
As described in `Node.js`' page, `import()` calls are allowed in `CommonJS` files:
https://nodejs.org/api/esm.html#import-expressions

Despite this fact, `webpack` replaces all `import()` calls with `require`s.

**How should this be implemented in your opinion?**
When targeting `commonjs2`, dynamic `import()` calls importing an external `module` should stay untouched. 

Furthermore, adding a `module` external should not report an error even if the target library type is `commonjs2`

**Are you willing to work on this yourself?**
no

**Comment by alexander-akait** [comment, marked as answer]
Do you looking for https://webpack.js.org/configuration/externals/#externalstype, i.e. `import`?

**Reply by neurolag** [reply]
Thanks for your rapid answer 😄 
I just gave it a go and just like the `module` type, `import` doesn't work either:

```
The target environment doesn't support 'import()' so it's not possible to use external type 'import'
```

So no, this is not what I have been looking for

**Reply by alexander-akait** [reply]
Do you want to use `import(...)` for externals only?

**Reply by neurolag** [reply]
> Do you want to use `import(...)` for externals only?

Yes
Executing `import(...)` statements on internal modules is working already, as far as I can tell

**Reply by alexander-akait** [reply]
Yes, it supported in Node.js, can you provide a real use case why you need this?

**Reply by neurolag** [reply]
I don't think this is a very common use case for `webpack`, but I'm using `webpack` to create a self-contained package for the use with `node` rather than a web browser.

The module itself must be `CommonJS` (which is why I use `commonjs2` as target library). I have a few `externals` modules (currently all `commonjs` modules) which are loaded from `node_modules` rather than from inside the bundle.

I find myself unable to update these `externals` modules because the new version of some of these packages are `"type": "module"` packages (aka. ESModule packages) which need to be `import()`ed rather than `require`d.

**Reply by alexander-akait** [reply]
sorry for delay, sounds like you can set `import(...)` supporing (https://github.com/webpack/webpack/blob/main/lib/ExternalModule.js#L139), can you try to set `output.environment.dynamicImport` to `true` and check again?

**Reply by neurolag** [reply]
> sorry for delay, sounds like you can set `import(...)` supporing (https://github.com/webpack/webpack/blob/main/lib/ExternalModule.js#L139), can you try to set `output.environment.dynamicImport` to `true` and check again?

No worries - thank you so much for helping me out 😄 
This did the trick! Thank you so much for helping me out 😄

**Comment by ntucker** [comment]
I find it odd that I have to write specific exception code every time I want to use a dynamic import

If I add import('@ant-design/cssinjs') now I need:

```js
  const ext = nodeExternals({
    allowlist: [/@babel\/runtime/],
    additionalModuleDirs: ['../../node_modules'],
  });
  config.externals = [
    function ({ context, request, contextInfo, getResolve }, callback) {
      if (request === '@ant-design/cssinjs') {
        callback(null, 'import ' + request);
      } else ext({ context, request, contextInfo, getResolve }, callback);
    },
  ];
```

Then if I decide to add another one, I'll have to add it to this list. Furthermore, that means that if I don't use a dynamic import it won't actually work as expected.

What's odd is that, if the environment supports dynamicImport - there's no reason to ever transform it. Why do the extra work and force everyone to do super hyper-specific configurations? Just leave it alone and everything works fine.

**Reply by alexander-akait** [reply]
Sorry for delay

> What's odd is that, if the environment supports dynamicImport - there's no reason to ever transform it.

To bundle them - move to chunks/merge/remove unsed/tree shakes/etc, i.e. optimixations, your solution is valid if you want to keep dynamic imports as is

**Reply by ntucker** [reply]
I'm having trouble comprehending this answer. I do *NOT* want to have to specify every library that I will dynamically import in my webpack configuration as I have already specified it in the code as a dynamic import that should not be transformed. Of course my solution is 'valid' in the sense that it works, but not in the sense that it makes 1) any sense 2) is scalable in any way 3) provides a good enough user experience that will prevent people from migrating away from webpack

**Reply by alexander-akait** [reply]
> I have already specified it in the code as a dynamic import that should not be transformed

Why do you think so? There are a lot of optimizations async modules via `import(...)`, so you just lose them

Other bundlers do the same. Also you can have multiple `@ant-design/cssinjs` impots - dynamic and async, what we should do here if we should leave `import` as is? I just want to say the function notation provide the real power. If you need more arguments in callback for checks just provide infromation what will help you

**Reply by neurolag** [reply]
> > I have already specified it in the code as a dynamic import that should not be transformed
> 
> Why do you think so? There are a lot of optimizations async modules via `import(...)`, so you just lose them
> 
> Other bundlers do the same. Also you can have multiple `@ant-design/cssinjs` impots - dynamic and async, what we should do here if we should leave `import` as is? I just want to say the function notation provide the real power. If you need more arguments in callback for checks just provide infromation what will help you

I have the impression that @ntucker has the exact same question that I have in mind right now:
Is webpack able to leave certain module inclusions as-is?

Consider I have the following `.js` code:

```js
import { MyClass } from "./MyClass.js";
import { createRequire } from "node:module";
import rescape from "@stdlib/utils-escape-regexp-string"; // this is a commonjs module

const path = createRequire(new URL(".", import.meta.url))("node:path"); // explicitly include `node:path` using the `require`-mechanism
rescape(`${new MyClass()}`);
(await import("inquirer")).default([]);
```

Consider I want `inquirer`, `node:module`, `@stdlib/utils-escape/regexp-string` and `node:path` to be externals.
Is there some way for me to specify that `webpack` shouldn't touch the expressions I am using for including the modules?`

Meaning:
  - `./MyClass.js` should be bundled
  - `node:module` should be `import`ed
  - `@stdlib/utils-escape-regexp-string` should be `import`ed (commonjs are treated as if they were ESModules with `default` exports)
  - `node:path` should be `require`d using `createRequire`
  - `inquirer` should be dynamically imported

Is there some way for me to reach this goal without having to specify the nature of the inclusion expression for each module?

**Reply by ntucker** [reply]
@alexander-akait 

I think you're assuming a web application target? Not library and not nodejs? If webpack should not support those, there shouldn't be [multiple ](https://webpack.js.org/configuration/externals/) [docs ](https://webpack.js.org/guides/author-libraries/) [pages ](https://webpack.js.org/configuration/target/#string) on them.

Otherwise, your comments don't make sense to me. 

> There are a lot of optimizations async modules via import(...), so you just lose them

Node: I don't consider hoisting a require() on an optional package to top level being an optimization. This means the first import will fail and *nothing* will work. It is in fact breaking, and defeats the entire purpose of using a dynamic import to begin with. Additionally, forcing parsing of all javascript up-front is an anti-optimization and increases startup times and memory usage.

Library: Diabling the ability of library consumers to use these optimizations *DISABLES* optimizations. So it is in fact better to keep these in place so they *can* be optimized properly.

> Other bundlers do the same

No rollup works fine and doesn't break this. However, I'd like to use webpack for everything as it has a better ecosystem and I can share things with my application bundler.

> what we should do here if we should leave import as is

I don't understand the nature of the question - you answered it yourself: 'leave import as is'. Obviously I'm not implying we should do this for web application bundling if that's what you're implying. [Externals](https://webpack.js.org/configuration/externals/) are definitely a thing in webpack.

> I just want to say the function notation provide the real power. If you need more arguments in callback for checks just provide infromation what will help you

Can you elaborate on what this is? Maybe there are docs links that explain more? I don't know what 'function notation' is in this context.

**Reply by alexander-akait** [reply]
@manuth @ntucker 

> Can you elaborate on what this is? Maybe there are docs links that explain more? I don't know what 'function notation' is in this context.

Using `function` to describe cases and please. And please be more calm, because I get the feeling that you are not trying to constructively talk, but to prove to me that I am a fool and wrong. 

Then let's go to the list of current problems point by point, because I see a lot of different things here, and it's hard to answer what exactly this could be achieved and how better. Can you summarize in table - what do you have and how it should be load externally.

We have a wide range of opportunities for externals and it's pretty hard to find something in common for everyone, I can put a lot of links here with incredibly different requests and statements that this is how it should work.

I’m not saying that we can’t improve it, I just want to reach an understanding of what exactly is broken at the moment or it’s not possible to  configure it without problem. So can we summarize, many things already support, so maybe we can just do better defaults or provide an option to enable such defaults

**Reply by neurolag** [reply]
> @manuth @ntucker
> 
> > Can you elaborate on what this is? Maybe there are docs links that explain more? I don't know what 'function notation' is in this context.
> 
> Using `function` to describe cases and please. And please be more calm, because I get the feeling that you are not trying to constructively talk, but to prove to me that I am a fool and wrong.
> 
> Then let's go to the list of current problems point by point, because I see a lot of different things here, and it's hard to answer what exactly this could be achieved and how better. Can you summarize in table - what do you have and how it should be load externally.
> 
> We have a wide range of opportunities for externals and it's pretty hard to find something in common for everyone, I can put a lot of links here with incredibly different requests and statements that this is how it should work.
> 
> I’m not saying that we can’t improve it, I just want to reach an understanding of what exactly is broken at the moment or it’s not possible to configure it without problem. So can we summarize, many things already support, so maybe we can just do better defaults or provide an option to enable such defaults

Thank you so much for hearing me out - I'll write a separate answer in this discussion as I don't think me and @ntucker are talking about the same thing.

**Comment by neurolag** [comment]
Is there some way to flag a dependency as `external` without overwriting its `externalType`?

Consider the following code:

```ts
import { createRequire } from "node:module";
import puppeteer from "puppeteer"; // I want to import the ESM version of `puppeteer` here

const anotherPuppeteer = createRequire(new URL(".", import.meta.url)("puppeteer"); // At some other place - in another source file, by chance, I want to import the CommonJS version of `puppeteer`
```

Is there some way to tell webpack that `puppeteer` is externl and that the statements should be left untouched? (meaning the `import` should stay an `import`, the `require`-statement should stay a `require`-statement)

That would be a great feature as you probably want to import the module exactly as you stated it in your code.

**Reply by alexander-akait** [reply]
I'm afraid not, do you want only certain packages to be untouched? Or every `import ... from ...`? Also what is your final output type - `library` or `application`?

**Reply by neurolag** [reply]
Thanks for coming back at me!
My final output type would be an npm module - so `library`, I guess.
Thus, I'd like to have any npm dependencies located under `node_modules` to be external and to have full control over what kind of inclusion statement is emitted (as in keep the statements as they were).

But I think webpack is rather focused on bundling apps for the web - as its included in the name.
So I'm totally fine if there's no way of getting this behaviour realized.

**Reply by felipellrocha** [reply]
I would like to have full control over which packages are left untouched, yeah, and I am targetting a web application. I don't see why this couldn't be realized.
